### DISCUSSION TITLE
ESP32 v12.5.0 json seems broken

### DISCUSSION THREAD

**Discussion by hansij66** [arendst/Tasmota#18817] [Q&A]
### PROBLEM DESCRIPTION
Tasmo admin reports broken json
https://github.com/TasmoAdmin/TasmoAdmin/issues/887
https://github.com/TasmoAdmin/TasmoAdmin/issues/893
https://github.com/TasmoAdmin/TasmoAdmin/issues/910

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [ ] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [ ] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [ ] Device used (e.g., Sonoff Basic): Shelly Plus 2pm
- [x] Tasmota binary firmware version number used: 12.5.0
  - [x] Pre-compiled
  - [ ] Self-compiled
- [x] Flashing tools used: RS232 + Chrome
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_

See Tasmo Admin issues

JSON ERROR => 4: Syntax error
Please copy the whole error message and copy it in a new issue on GitHub. Somehow the Tasmota Firmware answered a wrong formatted JSON string which needs to get adjusted in TasmoAdmin.
JSON Answer => {"Status":{"Module":0,"DeviceName":"Screen-studioludy-zijkant","FriendlyName":["Tasmota",""],"Topic":"tasmota_CE1EB8","ButtonTopic":"0","Power":2,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":1,"InfoRetain":0,"StateRetain":1,"StatusRetain":1}} (retained,"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota32/release/tasmota32solo1.bin","RestartReason":"Vbat power on reset","Uptime":"0T07:42:55","StartupUTC":"2023-05-14T10:31:49","Sleep":50,"CfgHolder":4617,"BootCount":7,"BCResetTime":"2023-05-06T12:53:20","SaveCount":25}} (retained,"StatusFWR":{"Version":"12.5.0(solo1)single-core","BuildDateTime":"2023-04-17T08:07:06","Core":"2_0_7","SDK":"4.4.4.20230218","CpuFrequency":240,"Hardware":"ESP32-U4WDH-S rev.3","CR":"473/699"}} (retained,"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["**********-lan",""],"TelePeriod":60,"Resolution":"558180C0","SetOption":["0000C029","2805C80001000600003C5A0A192800000000","00000080","00006000","80004080","00000000"]}} (retained,"StatusMEM":{"ProgramSize":1518,"Free":1361,"Heap":164,"StackLowMark":4,"PsrMax":0,"PsrFree":0,"ProgramFlashSize":4096,"FlashSize":4096,"FlashChipId":"164020","FlashFrequency":40,"FlashMode":"DIO","Features":["00000809","9F9AD7CF","0015A001","B7F7BFCF","05DA9BC0","E0360DC7","480840D2","20200000","D4BC482D","000A80B1"],"Drivers":"1,2,3,4,5,7,8,9,10,11,12,14,16,17,20,21,24,26,27,29,34,35,38,50,52,59,60,62,63,66,67,82,86,87,88","Sensors":"1,2,3,5,6,7,8,9,10,11,12,13,14,15,17,18,19,20,21,22,26,31,34,37,39,40,42,43,45,51,52,55,56,58,59,64,66,67,74,85,92,95,98,103,105,127","I2CDriver":"7,8,9,10,11,12,13,14,15,17,18,20,24,29,31,36,41,42,44,46,48,58,62,65,69,76,77"}} (retained,"StatusNET":{"Hostname":"tasmota-******-7864","IPAddress":"192.168.1.216","Gateway":"192.168.1.254","Subnetmask":"255.255.255.0","DNSServer1":"192.168.1.254","DNSServer2":"***********","Mac":"************","IP6Global":"************","IP6Local":"fe80::4617:93ff:fece:1eb8","Ethernet":{"Hostname":"","IPAddress":"0.0.0.0","Gateway":"0.0.0.0","Subnetmask":"0.0.0.0","DNSServer1":"192.168.1.254","DNSServer2":"******************","Mac":"00:00:00:00:00:00","IP6Global":"","IP6Local":""},"Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}} (retained,"StatusMQT":{"MqttHost":"192.168.1.1","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_CE1EB8","MqttUser":"********","MqttCount":6,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}} (retained,"StatusTIM":{"UTC":"2023-05-14T18:14:44","Local":"2023-05-14T20:14:44","StartDST":"2023-03-26T02:00:00","EndDST":"2023-10-29T03:00:00","Timezone":99,"Sunrise":"06:09","Sunset":"21:22"}} (retained,"StatusPTH":{"PowerDelta":[0,0,0,0,0,0,0,0],"PowerLow":0,"PowerHigh":0,"VoltageLow":0,"VoltageHigh":0,"CurrentLow":0,"CurrentHigh":0}} (retained,"StatusSNS":{"Time":"2023-05-14T20:14:44","Switch1":"OFF","Switch2":"OFF","ANALOG":{"Temperature1":21.7},"ENERGY":{"TotalStartTime":"1970-01-01T00:00:00","Total":0.000,"Yesterday":0.000,"Today":0.000,"Power":[0,0],"ApparentPower":[0,0],"ReactivePower":[0,0],"Factor":[0.00,0.00],"Frequency":50,"Voltage":237,"Current":[0.000,0.000]},"TempUnit":"C"}} (retained,"StatusSTS":{"Time":"2023-05-14T20:14:44","Uptime":"0T07:42:55","UptimeSec":27775,"Heap":163,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":6,"Berry":{"HeapUsed":3,"Objects":44},"POWER1":"OFF","POWER2":"ON","Wifi":{"AP":1,"SSId":"********-lan","BSSId":"18:E8:29:FB:96:4C","Channel":11,"Mode":"11n","RSSI":94,"Signal":-53,"LinkCount":2,"Downtime":"0T00:00:28"}}} (retained}

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._
Looks to be in the Status section of the payload.

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by Jason2866** [comment]
Please specify exactly what should be wrong. Since you posted the log not as code snipplet it is impossible to check.

**Comment by barbudor** [comment]
And try with latest development version
I see no problems in the JSON on my test versions

**Comment by s-hadinger** [comment]
JSON returned by Tasmota is valid. The logs above seem to include stuff added by some external stuff, namely `(retained, ` in between JSONs. Please refer to TasmotAdmin as we don't support this tool

**Reply by sfromis** [reply]
It is unclear how exact the copy/paste in those TasmoAdmin threads was done, but it seems that TasmoAdmin is not able to handle that the web response to the status command(s) will include the same as when commands are issued in the console, meaning ending with ` (retained)` when configuring with the `StatusRetain` option. Hence the JSON is broken if assuming that the response is "pure" JSON, and as Tasmota has no "actual API" stipulating a JSON-only response, tools doing web scraping always risks things not working any longer.

No such issues when doing it the proper way, using MQTT, where the response is just the JSON, as expected.

**Reply by barbudor** [reply]
The response to the web command is perfect on my 12.5.0.x devices
```
{
    "Status": {
        "Module": 0,
        "DeviceName": "Prise PC-Desk",
        "FriendlyName": [
            "Prise PC-Desk"
        ],
        "Topic": "skt-pcdesk",
        "ButtonTopic": "",
        "Power": 1,
        "PowerOnState": 0,
        "LedState": 1,
        "LedMask": "FFFF",
        "SaveData": 4,
        "SaveState": 0,
        "SwitchTopic": "0",
        "SwitchMode": [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0
        ],
        "ButtonRetain": 0,
        "SwitchRetain": 0,
        "SensorRetain": 1,
        "PowerRetain": 1,
        "InfoRetain": 1,
        "StateRetain": 0,
        "StatusRetain": 0
    },
    "StatusPRM": {
        "Baudrate": 115200,
        "SerialConfig": "8N1",
        "GroupTopic": "alldevices",
        "OtaUrl": "http://server/tasmota/latest/tasmota-bhome.bin.gz",
        "RestartReason": "Software/System restart",
        "Uptime": "18T13:12:36",
        "StartupUTC": "2023-05-20T06:47:59",
        "Sleep": 50,
        "CfgHolder": 4617,
        "BootCount": 61,
        "BCResetTime": "2021-05-28T13:47:56",
        "SaveCount": 951,
        "SaveAddress": "FB000"
    },
    "StatusFWR": {
        "Version": "12.5.0.1(bhome)",
        "BuildDateTime": "2023-04-20T00:01:20",
        "Boot": 31,
        "Core": "2_7_4_9",
        "SDK": "2.2.2-dev(38a443e)",
        "CpuFrequency": 80,
        "Hardware": "ESP8285N08",
        "CR": "468/699"
    },
    "StatusLOG": {
        "SerialLog": 2,
        "WebLog": 2,
        "MqttLog": 0,
        "SysLog": 0,
        "LogHost": "192.168.168.244",
        "LogPort": 5140,
        "SSId": [
            "*********",
            "*********"
        ],
        "TelePeriod": 300,
        "Resolution": "558980C0",
        "SetOption": [
            "00208238",
            "2805C80001000600003C5A18002800000000",
            "000000C4",
            "00006000",
            "09004100",
            "00000000"
        ]
    },
    "StatusMEM": {
        "ProgramSize": 565,
        "Free": 436,
        "Heap": 22,
        "ProgramFlashSize": 1024,
        "FlashSize": 1024,
        "FlashChipId": "144051",
        "FlashFrequency": 40,
        "FlashMode": "DOUT",
        "Features": [
            "00000809",
            "8F808587",
            "04000001",
            "000000CF",
            "000012C0",
            "40000901",
            "00024080",
            "00001000",
            "00000020",
            "00000000"
        ],
        "Drivers": "1,2,3,4,8,9,10,24,27,29,38,41,45",
        "Sensors": "1,2,3,5,6",
        "I2CDriver": ""
    },
    "StatusNET": {
        "Hostname": "skt-pcdesk",
        "IPAddress": "192.168.168.105",
        "Gateway": "192.168.168.254",
        "Subnetmask": "255.255.255.0",
        "DNSServer1": "192.168.168.244",
        "DNSServer2": "1.1.1.1",
        "Mac": "24:62:AB:2F:93:FC",
        "Webserver": 2,
        "HTTP_API": 1,
        "WifiConfig": 5,
        "WifiPower": 17.0
    },
    "StatusMQT": {
        "MqttHost": "192.168.168.244",
        "MqttPort": 1883,
        "MqttClientMask": "%012X",
        "MqttClient": "2462AB2F93FC",
        "MqttUser": "skt-pcdesk",
        "MqttCount": 3,
        "MAX_PACKET_SIZE": 1200,
        "KEEPALIVE": 30,
        "SOCKET_TIMEOUT": 4
    },
    "StatusTIM": {
        "UTC": "2023-06-07T20:00:35",
        "Local": "2023-06-07T22:00:35",
        "StartDST": "2023-03-26T02:00:00",
        "EndDST": "2023-10-29T03:00:00",
        "Timezone": 99,
        "Sunrise": "05:48",
        "Sunset": "21:49"
    },
    "StatusPTH": {
        "PowerDelta": [
            0,
            0,
            0
        ],
        "PowerLow": 0,
        "PowerHigh": 0,
        "VoltageLow": 0,
        "VoltageHigh": 0,
        "CurrentLow": 0,
        "CurrentHigh": 0
    },
    "StatusSNS": {
        "Time": "2023-06-07T22:00:35+02:00",
        "ENERGY": {
            "TotalStartTime": "2021-05-28T13:47:56",
            "Total": 13.807,
            "Yesterday": 0.061,
            "Today": 0.016,
            "Power": 27.6,
            "ApparentPower": 32.6,
            "ReactivePower": 17.3,
            "Factor": 0.85,
            "Voltage": 245,
            "Current": 0.133
        }
    },
    "StatusSTS": {
        "Time": "2023-06-07T22:00:35+02:00",
        "Uptime": "18T13:12:36",
        "UptimeSec": 1602756,
        "Heap": 24,
        "SleepMode": "Dynamic",
        "Sleep": 50,
        "LoadAvg": 1713,
        "MqttCount": 3,
        "POWER": "ON",
        "Wifi": {
            "AP": 1,
            "SSId": "********",
            "BSSId": "88:C3:97:F2:25:EC",
            "Channel": 1,
            "Mode": "11n",
            "RSSI": 54,
            "Signal": -73,
            "LinkCount": 1,
            "Downtime": "0T00:00:06"
        }
    }
}
```

**Reply by sfromis** [reply]
You are not running with the `StatusRetain` option on, @barbudor - that is what I saw make the difference.

**Reply by barbudor** [reply]
Ok, I think I've spotted it

**Reply by barbudor** [reply]
But I don't see how to solve that easily
The thing is that the response to a HTTP command is pulled from the log buffer
And the log buffer includes ` (retained)` at the end of each line is StatusRetain is ON and the command is a Status command

I would say : if you are using TasmoAdmin that is a HTTP API based tool, you are probably not using MQTT, so don't use StatusRetain ON.
If you are using MQTT, then don't use TasmoAdmin and switch to TDM

**Reply by sfromis** [reply]
Yeah, that's why I ranted a bit about the dangers of web scraping, instead of have an API with a stipulated return data format. Of course, TasmoAdmin could handle it by sanitizing the returned data, like only picking what's within outermost `{}`, or more crudely trim the end.

When it comes to HTTP vs MQTT, one place where people are led to see TasmoAdmin as a mainstream tool for Tasmota is when HA (which is very much MQTT-based with Tasmota), includes TasmoAdmin as preconfigured extension (or whatever the term is), for easy installation, and one click away after that. Thus, I'd expect a significant overlap of protocol usage.

**Reply by barbudor** [reply]
I have a fix but as it is really an edge case, you have to pray for Theo to accept it
https://github.com/arendst/Tasmota/pull/18819

**Reply by sfromis** [reply]
Hmmmm..... At least `InfoRetain`, `SensorRetain` and `PowerRetain` has a similar situation with the log output.
```
        "ButtonRetain": 0,
        "SwitchRetain": 0,
        "SensorRetain": 1,
        "PowerRetain": 1,
        "InfoRetain": 1,
        "StateRetain": 0,
        "StatusRetain": 0
```

Ok, they may not directly affect responses to http commands, but log scraping could still be hit.

Maybe a more general solution (including the option of "doing nothing") should be considered.

**Reply by barbudor** [reply]
I forgot about State, and may be Power too

All others 'retain' cases are, AFAIK, for MQTT message that are sent directly, not as answers to commands

To make it generic, may be search for '(retained)' at the end of the log which could be possible to as it is a fixed length from the end. 

I'll pause the PR for now

Edit : or searching for the last '}'

Edit2: Publish2 would also be concerned although it would really be a edge edge case to use a http command to publish on mqtt

**Reply by sfromis** [reply]
I did test, and found what could be called a "bug" (as in being inconsistent behavior), the `(retained)` part is not in the payload resulting from commands `State` or `TelePeriod`, hence this does not affect the http command response. And when it comes to Power, the `(retained)` part is globbed on to the non-JSON payload (not returned to http), but not to the JSON payload (returned to http) I saw after a Power command. Thus, for the specific case of http `cm?cmnd=`, I'm not sure that other than `Status` response is actually affected by sprinkling on retain options (which I'm *really* not a fan of anyway).

Still, for anyone doing other types of log scraping - like serial, MqttLog (as opposed to MQTT), Syslog or Berry - the `(retained)` part after some of the JSON payloads do have to be dealt with, to get to valid JSON for parsing. And yes, I have seen at least one case of a user going hard for automation via Syslog message parsing, instead of MQTT.

I'm inclined to think that people going for scraping have chosen to experience pain. An API with well-defined output would make sense to have, and here I do see an opportunity to deliberately split hairs.

Since the "Configure Other" page has explicitly designated that Tasmota has **HTTP API enable**, this constitutes a strong suggestion of it being reasonable to expect stringent predictable responses (like clean JSON) from http commands, meaning that the `(retained)` part constitutes polluting dirt which does not belong in API responses.

When it comes to Tasmota logging, it could also be called a "bug" (as in inconsistent behavior) that the `(retained)` indicator is not present in all messages about MQTT retained publish, but I'm not arguing in favor of ensuring this :grinning: 

To be clear, while I'm no fan of having those retain configuration options, I do see that there are use cases helped by this being possible, as in clients always being able to get information when subscribing, without having to issue commands to query setup and sensor details, which may not always be supported.

**Reply by s-hadinger** [reply]
The redline that should not be crossed, is that HTTP API should return valid JSON without any decorator like `(retained)`.

Do I understand that some commands (Power?) add `(retained)` to the HTTP payload response? If we really want the retain flag, it could be added as a header, although I don't see any value of mentioning it at all.

**Reply by sfromis** [reply]
The specific example is that the `Status` command outputs a payload with the sometimes rather long JSON, followed by the `(retained)` tag if the user has set the option `StatusRetain`.

In principle, the log output for any command will have this tag, if the payload is published as retained, but that's not consistently the case, like when (as already mentioned), the `State` command will not have this included in the response to a http request, even when having `StateRetain` set, and the Power command output (with `PowerRetain`) only has the `(retained)` tag on the non-JSON log message for this MQTT publish, but not on the JSON response to the http request, and not in the log for this either (yes that's two messages in the log).

Tasmota is not at all consistent about log messages for retained MQTT payloads being tagged with `(retained)`, and I'm not specifically aware of other commands than `Status` where that tag is included in the http `cm?cmnd=` response. However, people wanting to parse Tasmota logging by other means than http responses still has to suffer with the tag for some log messages which otherwise has a "more clean" JSON format, if they activate the associated `xxxxxRetain` option.

**Reply by barbudor** [reply]
Just reworked the PR
Simpler, nicer and works in every cases
Tested with different variants of Status but also with State, Teleperiod Power with both StatusRetain ON and OFF
Also test with publish and publish2

**Reply by barbudor** [reply]
@s-hadinger 
The reason why "(retained)" appear in the HTTP response is that the HTTP response is "peeked" from the log buffer. And any MQTT publish create a log entry where " (retained)" is appended in the log if hte publish is with retain.
So it's not the command that adds "retained" but the MQTT driver adds it to the log

I have now removed it, parsing the log line backwards until closing brace.

**Comment by Jason2866** [comment]
My vote is for no change in Tasmota. Solution do not use "retain" It is a bad idea at all.
TasmoAdmin has to take care of his weird use of http.

**Reply by s-hadinger** [reply]
I vote for a fix ;)

Matter remote HTTP requests needs clean JSON responses.

**Reply by barbudor** [reply]
My fix merged by Theo this morning
